hummingbot/hummingbot
# feat/gate io connector update

**Before submitting this PR, please make sure**:


**A description of the changes proposed in the pull request**:
GateIO exchance connector refactoring to bring the code to the latest standards. 
The class ExchangePyBase was modified to serve as the new building base for centralized exchanges connectors. All the generic logic was moved to it and also the methods expected to be implemented by subclasses were created as abstract methods. The following connectors have been modified to be now subclasses of the new base class:
- BInance spot
- Kucoin
- GateIO
Proposal of `integration_test_tool.py` in `test/connector` to aid manual testing. The script uses real connections with the exchanges and executes basic operations without mocking anything.


**Tests performed by the developer**:
Unit tests for Binance, Kucoin and GateIO
Live interactive testing for gate io
Pure mm for gate io with spreads of 0.5% and 10s refresh time


**Test summary**:
- Install and start the feature branch without any issue
- Connect Binance, Gate_io, and KuCoin API key. Gate_io did have an issue with accepting invalid API keys which had been fixed within the commits.
- Checked the balance and successfully returns and displays accurate balance in each exchange.
- Created different pure market-making strategies using Binance, Gate_io, and KuCoin as exchanges. A minor issue was seen during the strategy creation specific to Gate_io which has been sorted within the commits.
- Started the strategy, initially, errors occurred when starting any strategy which was confirmed fixed in the latest commit.
- Confirmed that the order creation and cancellation follow the specified parameter set in the strategy.
- Stopped and restarted the strategy multiple times to verify that orders are properly getting canceled.
- Tested on different refresh rates to confirm that there are no orders left out in the exchange.
- Filled orders are confirmed and recorded correctly and fees match the fee record in the exchange.
- Checked the orders and prices to make sure that they are synchronized with the exchange.
- Run multiple bots using the same exchange and market to confirm that there are no orderID conflicts.
- Add the API key by creating a python file and testing the integration_test_tool.py. Confirms that the tool is running and capturing issues if there are any.


**Tips for QA testing**:
Please review Binance, Kucoin and GateIO connectors.
If you want to use the new script to do the basic functionality test it can be run with: 
```
conda activate hummingbot
PYTHONPATH=. python test/connector/integration_test_tool.py
```
(press enter to see the prompt)


This PR comes from https://github.com/CoinAlpha/hummingbot/pull/198
